src/Lib.P2P/Cryptography/Key.cs

```diff
@@ -0,0 +1,187 @@
+        }
+
+        /// <summary>
+        ///   Create a signature for the data.
+        /// </summary>
+        /// <param name="data">
+        ///   The data to sign.
+        /// </param>
+        /// <returns>
+        ///   The signature.
+        /// </returns>
+        public byte[] Sign(byte[] data)
+        {
+            var signer = SignerUtilities.GetSigner(_signingAlgorithmName);
+            signer.Init(true, _privateKey);
+            signer.BlockUpdate(data, 0, data.Length);
+            return signer.GenerateSignature();
+        }
+
+        /// <summary>
+        ///   Create a public key from the IPFS message.
+        /// </summary>
+        /// <param name="bytes">
+        ///   The IPFS encoded protobuf PublicKey message.
+        /// </param>
+        /// <returns>
+        ///   The public key.
+        /// </returns>
+        public static Key CreatePublicKeyFromIpfs(byte[] bytes)
+        {
+            var key = new Key();
+
+            var ms = new MemoryStream(bytes, false);
+            var ipfsKey = Serializer.Deserialize<PublicKeyMessage>(ms);
+
+            switch (ipfsKey.Type)
+            {
+                case KeyType.Rsa:
+                    key._publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key._signingAlgorithmName = RsaSigningAlgorithmName;
+                    break;
+                case KeyType.Ed25519:
+                    key._publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key._signingAlgorithmName = Ed25519SigningAlgorithmName;
+                    break;
+                case KeyType.Secp256K1:
+                    key._publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key._signingAlgorithmName = EcSigningAlgorithmName;
+                    break;
+                default:
+                    throw new InvalidDataException($"Unknown key type of {ipfsKey.Type}.");
+            }
+
+            return key;
+        }
+
+        /// <summary>
+        ///   Create the key from the Bouncy Castle private key.
+        /// </summary>
+        /// <param name="privateKey">
+        ///   The Bouncy Castle private key.
+        /// </param>
+        public static Key CreatePrivateKey(AsymmetricKeyParameter privateKey)
+        {
+            var key = new Key();
+            key._privateKey = privateKey;
+
+            // Get the public key from the private key.
+            if (privateKey is RsaPrivateCrtKeyParameters rsa)
+            {
+                key._publicKey = new RsaKeyParameters(false, rsa.Modulus, rsa.PublicExponent);
+                key._signingAlgorithmName = RsaSigningAlgorithmName;
+            }
+            else if (privateKey is Ed25519PrivateKeyParameters ed)
+            {
+                key._publicKey = ed.GeneratePublicKey();
+                key._signingAlgorithmName = Ed25519SigningAlgorithmName;
+            }
+            else if (privateKey is ECPrivateKeyParameters ec)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add the missing 'else' clause.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)